lib/adt.ex

```diff
@@ -24,7 +24,38 @@ defmodule ADT do
       end
     end
     modules = values |> Enum.map(&generate_defmodule/1)
-    Enum.concat([[variants_definition], modules, [variants_reader]])
+    Enum.concat([[variants_definition], modules, [variants_reader, case_macro(values)]])
+  end
+
+  defp case_macro(values) do
+    quote do
+      defmacro case(a, [do: statements]) do
+        possible_variants = Enum.map(
+          unquote(values),
+          fn {variant, _} ->
+            # Maps a module like Foo.Bar.Baz into a short string "Baz"
+            Regex.named_captures(~r/\.(?<short>[^\.]+)$/, inspect(variant), include_captures: true) |> Map.fetch!("short")
+          end
+        ) |> Enum.sort
+        given_variants = statements |> Enum.map(fn {k, _} -> to_string(k) end) |> Enum.sort
```

Comment:
seeing as we need `k` to be a string pretty much everywhere, would it make more sense to map it when we get it?